File: IntegrationTool/IntegrationTool/Controllers/IntegrationController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using System.Globalization;
using Newtonsoft.Json;
using IntegrationTool.Models;
using ClassLibrary;
using System.Text.RegularExpressions;

namespace IntegrationTool.Controllers
{
    [Authorize]
    public class IntegrationController : Controller
    {
        private IntegrationModel integrationConfigurationModel;
        private Encrypt encryptor;

        // ================================================================================================================
        // Retornar las vitas del sistema.
        // ================================================================================================================
        [HttpGet]
        public ActionResult configurationScheduled()
        {
            return View();
        }

        [HttpGet]
        public ActionResult configurationManual()
        {
            return View();
        }

        [HttpGet]
        public ActionResult manual()
        {
            return View();
        }

        [HttpGet]
        public ActionResult calendar()
        {
            return View();
        }

        // ================================================================================================================
        // Almacenar en la base de datos integraciones automaticas y manuales.
        // ================================================================================================================
        [HttpPut]
        public void saveIntegrationManual()
        {
            string resp = "";
            try
            {
                List<QueryParameter> queryParameters = getQueryParameters(Request);

                connectModel();

                int integrationId = integrationConfigurationModel.saveIntegrationManual(Convert.ToInt32(Request.Form["UserId"]),
                                                            Request.Form["IntegrationName"],
                                                            Convert.ToInt32(Request.Form["WebServiceId"]),
                                                            Convert.ToInt32(Request.Form["DatabaseParametersId"]),
                                                            Convert.ToInt32(Request.Form["FlatFileParameterId"]),
                                                            Convert.ToInt32(Request.Form["IntegrationTypeId"]),
                                                            Convert.ToInt32(Request.Form["QueryId"]),
                                                            Convert.ToInt32(Request.Form["OperationWebServiceId"]),
                                                            queryParameters,
                                                            Request.Form["CurlParameters"]);

                executeIntegrationManual(integrationId);

                resp = "{\"type\":\"success\", \"message\":\"Integration successful stored.\"}";
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Integration unsuccessful stored. Please try again.\"}";
            }


            response(resp);
        }

        [HttpPut]
        public void saveIntegrationScheduled()
        {
            string resp = "";
            try
            {
                List<QueryParameter> queryParameters = getQueryParameters(Request);

                connectModel();

                string format = "ddd MMM dd yyyy HH:mm:ss 'GMT'K '(Central America Standard Time)'";

                DateTime executionStartDate = DateTime.ParseExact(Request.Form["ExecutionStartDate"], format, CultureInfo.CurrentCulture);
                DateTime executionEndDate = DateTime.ParseExact(Request.Form["ExecutionEndDate"], format, CultureInfo.CurrentCulture);

                integrationConfigurationModel.saveIntegrationSchedule(Convert.ToInt32(Request.Form["UserId"]),
                                                            Request.Form["IntegrationName"],
                                                            Convert.ToInt32(Request.Form["WebServiceId"]),
                                                            Convert.ToInt32(Request.Form["DatabaseParametersId"]),
                                                            Convert.ToInt32(Request.Form["FlatFileParameterId"]),
                                                            Convert.ToInt32(Request.Form["IntegrationTypeId"]),
                                                            Convert.ToInt32(Request.Form["QueryId"]),
                                                            Convert.ToInt32(Request.Form["OperationWebServiceId"]),
                                                            queryParameters,
                                                            executionStartDate,
                                                            executionEndDate,
                                                            Convert.ToInt32(Request.Form["RecurrenceId"]),
                                                            Convert.ToInt32(Request.Form["StatusId"]),
                                                            Request.Form["Emails"],
                                                            Request.Form["CurlParameters"]);

                resp = "{\"type\":\"success\", \"message\":\"Integration successful stored.\"}";
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Integration unsuccessful stored. Please try again.\"}";
            }


            response(resp);
        }

        // ================================================================================================================
        // Actualizar en la base de datos integraciones automaticas.
        // ================================================================================================================
        [HttpPost]
        public void updateIntegrationScheduled()
        {
            string resp = "";
            try
            {
                List<QueryParameter> queryParameters = getQueryParameters(Request);

                connectModel();

                string format = "ddd MMM dd yyyy HH:mm:ss 'GMT'K '(Central America Standard Time)'";

                DateTime executionStartDate = DateTime.ParseExact(Request.Form["ExecutionStartDate"], format, CultureInfo.CurrentCulture);
                DateTime executionEndDate = DateTime.ParseExact(Request.Form["ExecutionEndDate"], format, CultureInfo.CurrentCulture);

                integrationConfigurationModel.updateIntegrationSchedule(Convert.ToInt32(Request.Form["IntegrationId"]),
                                                            Convert.ToInt32(Request.Form["UserId"]),
                                                            Request.Form["IntegrationName"],
                                                            Convert.ToInt32(Request.Form["WebServiceId"]),
                                                            Convert.ToInt32(Request.Form["DatabaseParametersId"]),
                                                            Convert.ToInt32(Request.Form["FlatFileParameterId"]),
                                                            Convert.ToInt32(Request.Form["IntegrationTypeId"]),
                                                            Convert.ToInt32(Request.Form["QueryId"]),
                                                            Convert.ToInt32(Request.Form["OperationWebServiceId"]),
                                                            queryParameters,
                                                            executionStartDate,
                                                            executionEndDate,
                                                            Convert.ToInt32(Request.Form["RecurrenceId"]),
                                                            Convert.ToInt32(Request.Form["StatusId"]),
                                                            Request.Form["Emails"],
                                                            Request.Form["CurlParameters"]);

                resp = "{\"type\":\"success\", \"message\":\"Integration successful updated.\"}";
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Integration unsuccessful updated. Please try again.\"}";
            }


            response(resp);
        }

        // ================================================================================================================
        // Obtiene y retorna datos necesarios para realizar una integración.
        // ================================================================================================================
        [HttpGet]
        public void getQueriesByIntegrationType(int id)
        {
            string resp = "";
            try
            {
                connectModel();
                List<Query> queries = integrationConfigurationModel.getQueriesByIntegrationType(id);

                foreach (Query query in queries)
                {
                    query.Query1 = encryptor.decryptData(query.Query1);
                    query.QueryName = encryptor.decryptData(query.QueryName);
                    query.Description = encryptor.decryptData(query.Description);
                }

                resp = serializeObject(queries);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the queries. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void getOperationsWebServices()
        {
            string resp = "";
            try
            {
                connectModel();
                List<OperationsWebService> operations = integrationConfigurationModel.getOperationsWebServices();

                resp = serializeObject(operations);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the operations web services. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void getIntegrationCategories()
        {
            string resp = "";
            try
            {
                connectModel();
                List<IntegrationCategory> integrationCategories = integrationConfigurationModel.getIntegrationCategories();

                resp = serializeObject(integrationCategories);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the integration categories. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void getRecurrences()
        {
            string resp = "";
            try
            {
                connectModel();
                List<Recurrence> recurrences = integrationConfigurationModel.getRecurrences();

                resp = serializeObject(recurrences);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the recurrences. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void getSheduleIntegrations()
        {
            string resp = "";
            try
            {
                connectModel();
                List<IntegrationTool.Models.Integration> integrations = integrationConfigurationModel.getScheduleIntegrations();

                resp = serializeObject(integrations);
            }
            catch (Exception)
            {              
               resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the schedule integration. Please try again.\"}";
            }

            response(resp);
        }


        [HttpGet]
        public void getManualIntegrations()
        {
            string resp = "";
            
            try
            {
                connectModel();
                List<IntegrationTool.Models.Integration> integrations = integrationConfigurationModel.getManualIntegrations();

                
                for (int index = 0; index < integrations.Count; index++ )
                {
                    try
                    {
                        integrations[index].WebService.Endpoint = encryptor.decryptData(integrations[index].WebService.Endpoint);
                    }
                    catch (Exception)
                    {
                        
                    }

                    try
                    {
                        integrations[index].DatabaseParameter.Name = encryptor.decryptData(integrations[index].DatabaseParameter.Name);
                    }
                    catch (Exception)
                    {

                    }
            
                    try
                    {
                        integrations[index].DatabaseParameter.Port = (integrations[index].DatabaseParameter.Port != null) ? encryptor.decryptData(integrations[index].DatabaseParameter.Port) : null;
                    }
                    catch (Exception)
                    {

                    }

                    try
                    {
                        integrations[index].DatabaseParameter.Instance = (integrations[index].DatabaseParameter.Instance != null) ? encryptor.decryptData(integrations[index].DatabaseParameter.Instance) : null;
                    }
                    catch (Exception)
                    {

                    }

                    try
                    {
                        integrations[index].DatabaseParameter.Instance = (integrations[index].DatabaseParameter.Instance != null) ? encryptor.decryptData(integrations[index].DatabaseParameter.Instance) : null;
                    }
                    catch (Exception)
                    {
                        
                    }

                    try
                    {
                        integrations[index].DatabaseParameter.Ip = encryptor.decryptData(integrations[index].DatabaseParameter.Ip);
                    }
                    catch (Exception)
                    {

                    }

                    try
                    {
                        integrations[index].FlatFilesParameter.Location = encryptor.decryptData(integrations[index].FlatFilesParameter.Location);
                    }
                    catch (Exception)
                    {

                    }

                    try
                    {
                        integrations[index].Query.Query1 = encryptor.decryptData(integrations[index].Query.Query1);
                    }
                    catch (Exception)
                    {

                    }
                }

                resp = serializeObject(integrations);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the manual integration. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void getIntegration(int id)
        {
            string resp = "";
            try
            {
                connectModel();
                IntegrationTool.Models.Integration integration = integrationConfigurationModel.getIntegration(id);

                resp = serializeObject(integration);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the integration selected. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void getStatus()
        {
            string resp = "";
            try
            {
                connectModel();
                IList<Status> status = integrationConfigurationModel.getStatus();

                resp = serializeObject(status);
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the integration selected. Please try again.\"}";
            }

            response(resp);
        }

        [HttpGet]
        public void executeIntegration(int id)
        {
            string resp = "";
            try
            {
                executeIntegrationManual(id);
                resp = "{\"type\":\"success\", \"message\":\"Integration successful executed.\"}";
            }
            catch (Exception)
            {
                resp = "{\"type\":\"danger\", \"message\":\"Can not be loaded the integration selected. Please try again.\"}";
            }

            response(resp);
        }

        // ================================================================================================================
        // Metodos privados que proveen funcionalidad a las acciones del controlador.
        // ================================================================================================================
        private void connectModel()
        {
            integrationConfigurationModel = new IntegrationModel();
            encryptor = new Encrypt();
        }

        private void response(string json)
        {
            Response.Clear();
            Response.ContentType = "application/json; charset=utf-8";
            Response.Write(json);
            Response.End();
        }

        private void executeIntegrationManual(int integrationId)
        {
            ClassLibrary.Integration integration = new ClassLibrary.Integration();
            integration.executeIntegration(integrationId);
        }

        private List<QueryParameter> getQueryParameters(HttpRequestBase req)
        {
            List<QueryParameter> queryParameters = new List<QueryParameter>();
            Regex regex = new Regex(@"\[\[.*\]\]");

            for (int index = 0; index < req.Form.Count; index++)
            {
                Match match = regex.Match(Request.Form.AllKeys[index]);
                if (match.Success)
                {
                    QueryParameter qp = new QueryParameter();
                    qp.Name = Request.Form.AllKeys[index];
                    string[] values = Request.Form[qp.Name].Split('|');

                    string value = "";
                    foreach(string v in values)
                    {
                        value += "'" +  v + "',";
                    }
                    qp.Value = value.TrimEnd(',');

                    queryParameters.Add(qp);
                }
            }

            return queryParameters;
        }

        private string serializeObject(Object obj)
        {
            return Newtonsoft.Json.JsonConvert.SerializeObject(obj,
                new JsonSerializerSettings()
                {
                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore
                });
        }
    }
}

File: IntegrationTool/IntegrationTool/Content/Scripts/Controllers/IntegrationController.js
var IntegrationController = function ($scope, $http, $routeParams) {
    $scope.typeMessage = 0;
    $scope.message = "";
    $scope.request = {};
    $scope.integrationsType = [];
    $scope.queries = [];
    $scope.webServices = [];
    $scope.operationsWebServices = [];
    $scope.databases = [];
    $scope.flatFiles = [];
    $scope.integrationCategories = [];
    $scope.recurrences = [];
    $scope.params = [];
    $scope.query;
    $scope.database;
    $scope.sendRequest = sendRequest;
    $scope.getQueries = getQueries;
    $scope.getQuery = getQuery;
    $scope.getRecurrences = getRecurrences;
    $scope.getDatabase = getDatabase;

    getIntegrationsType();
    getWebServices();
    getOperationsWebServices();
    getDatabases();
    getFlatFiles();
    getIntegrationCategories();

    if ($routeParams.id != -1) {
        getIntegration($routeParams.id);
    }

    function integration(resp) {
        resp.IntegrationTypeId = resp.IntegrationTypeId + '';
        resp.IntegrationCategoryId = resp.IntegrationCategoryId + '';
        resp.WebServiceId = resp.WebServiceId + '';
        resp.OperationWebServiceId = resp.OperationWebServiceId + '';
        resp.DatabaseParametersId = resp.DatabaseParametersId + '';
        resp.FlatFileParameterId = resp.FlatFileParameterId + '';
        resp.QueryId = resp.QueryId + '';

        getDatabase(resp.DatabaseParametersId);
        getQueries(resp.IntegrationTypeId);
        var Interval = setInterval(function () {
            if ($scope.queries.length != 0) {
                clearInterval(Interval);
                getQuery(resp.QueryId);
            }
        }, 50);

        if (resp.IntegrationCategoryId == 2) {
            getRecurrences(resp.IntegrationCategoryId);
            resp.RecurrenceId = resp.Calendars[0].RecurrenceId + '';
            resp.Emails = resp.Calendars[0].Emails + '';
            var startDate = new Date(resp.Calendars[0].ExecutionStartDate);
            var endDate = new Date(resp.Calendars[0].ExecutionEndDate);
            startDate.setHours(startDate.getHours() + 6);
            endDate.setHours(endDate.getHours() + 6);
            resp.ExecutionStartDate = startDate;
            resp.ExecutionEndDate = endDate;
        }

        $scope.request = resp;

        for (index = 0; index < resp.QueryParameters.length; index++) {
            $scope.request[resp.QueryParameters[index].Name] = resp.QueryParameters[index].Value.replace(/\'/g, '');
        }
    }

    function getIntegration(id) {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Integration/getIntegration?id=' + id, data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                integration(resp);
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getDatabase(id) {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Configuration/getDataBase?id=' + id, data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.database = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function sendRequest(req, form) {
        if (!form.$valid) {
            $scope.message = "Error: Invalid form, please try again.";
            $scope.typeMessage = "danger";
            return false;
        }

        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        console.log(req);

        var data = $.param(req);

        $http.put('Integration/saveIntegration', data, config).success(function (resp) {
            $scope.message = resp.message;
            $scope.typeMessage = resp.type;
            $scope.request = {};
            $scope.queries = [];
            $scope.recurrences = [];
            $scope.params = [];
            $scope.query = null;
            form.$setPristine();
            form.$setUntouched();
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getIntegrationsType() {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Configuration/getIntegrationsType', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.integrationsType = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getQueries(integrationTypeId) {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Integration/getQueriesByIntegrationType?id=' + integrationTypeId, data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.queries = resp;
                $scope.query = null;
                $scope.params = [];
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getWebServices() {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Configuration/getListWebServices', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.webServices = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getOperationsWebServices() {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Integration/getOperationsWebServices', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.operationsWebServices = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getQuery(queryId) {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Configuration/getQuery?id=' + queryId, data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.query = resp;
                getParams();
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getParams() {
        params = $scope.query.Query1.match(/\[\[.*\]\]/g);
        requestParams = [];
        for (index = 0; index < params.length; index++) {
            property = 'param' + index;
            requestParams[index] = { nameParam: params[index], valueParam: '' }
        }

       
        $scope.params = requestParams;
    }

    function getDatabases() {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Configuration/getListDatabases', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.databases = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getFlatFiles() {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Configuration/getListFlatFiles', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.flatFiles = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getIntegrationCategories() {
        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Integration/getIntegrationCategories', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.integrationCategories = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }

    function getRecurrences(integrationCategory) {
        if (integrationCategory == 1) {
            $scope.recurrences = [];
            return;
        }

        var config = {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8;'
            }
        }

        var data = $.param({});

        $http.get('Integration/getRecurrences', data, config).success(function (resp) {
            if (resp.type !== 'danger') {
                $scope.recurrences = resp;
            } else {
                $scope.message = resp.message;
                $scope.typeMessage = resp.type;
            }
        }).error(function (resp) {
            $scope.message = "Error: " + resp;
            $scope.typeMessage = "danger";
        });
    }
}

IntegrationController.$inject = ['$scope', '$http', '$routeParams'];